src/dmd/backend/cgcse.d

```diff
@@ -85,26 +124,128 @@ struct CSE
 
         if (alignsize >= 16 && TARGET_STACKALIGN >= 16)
         {
-            alignment = alignsize;
+            alignment_ = alignsize;
             STACKALIGN = alignsize;
             enforcealign = true;
         }
     }
-}
 
-// Returns: offset of slot i
-uint CSE_offset(int i)
-{
-    return i * CSE.slotSize;
-}
+    /****
+     * Get range of all CSEs filtered by matching `e`,
+     * starting with most recent.
+     * Returns:
+     *  input range
+     */
+    static auto filter(const elem* e)
+    {
+        struct Range
+        {
+            const elem* e;
+            int i;
 
-// Returns: true if CSE was ever loaded
-bool CSE_loaded(int i)
-{
-    return i < csextab.length &&   // array could be shrunk for non-CSEload entries
-           (csextab[i].flags & CSEload);
+          nothrow:
+
+            bool empty()
+            {
+                while (i)
+                {
+                    if (csextab[i - 1].e == e)
+                        return false;
+                    --i;
+                }
+                return true;
+            }
+
+            ref CSE front() { return csextab[i - 1]; }
+
+            void popFront() { --i; }
+        }
+
+        return Range(e, cast(int)csextab.length);
+    }
+
+    /*********************
+     * Remove instances of `e` from CSE table.
+     */
+    static void remove(const elem* e)
+    {
+        foreach (ref cse; csextab[])
+        {
+            if (cse.e == e)
+                cse.e = null;
+        }
+    }
+
+    /************************
+     * Create mask of registers from CSEs that refer to `e`.
+     * Returns:
+     *  mask
+     */
+    static regm_t mask(const elem* e)
+    {
+        regm_t result = 0;
+        foreach (ref cse; csextab[])
+        {
+            if (cse.e)
+                elem_debug(cse.e);
+            if (cse.e == e)
+                result |= cse.regm;
+        }
+        return result;
+    }
+
+    /***
+     * Finish generating code for this function.
+     * Get rid of unused cse temporaries by shrinking the array.
+     * References: loaded()
+     */
+    static void finish()
+    {
+        while (csextab.length != 0 && (csextab[csextab.length - 1].flags & CSEload) == 0)
+            csextab.setLength(csextab.length - 1);
+    }
+
+    /**** The rest of the functions can be called only after finish() ****/
+
+    /******************
+     * Returns:
+     *    total size used by CSE's
+     */
```

Comment:
I think you can use the more compact Ddoc in the form of `///` as has been done for `offset`.